odoo/fields.py

```diff
@@ -660,6 +660,37 @@ def get(key, get_context=env.context.get):
 
         return tuple(get(key) for key in self.depends_context)
 
+    def mapped(self, records):
+        if self.name == 'id':
+            # not stored in cache
+            return records._ids
+
+        if self.compute:
+            # a recompute must be forced for records to_recompute that intersect with input records
+            # otherwise an infinite loop will occur on the next call to flush() / recompute()
+            # as the value will be False in cache and will never trigger a MissingError which,
+            # in turn, triggers a recompute
+            to_compute_ids = records.env.all.tocompute.get(self, set())
+            if not to_compute_ids.isdisjoint(records._ids):
+                to_compute_ids = to_compute_ids.intersection(records._ids).difference(
+                    records.env._protected.get(self, ()))
+                self.compute_value(records.browse(to_compute_ids.intersection(records._ids)))
+
+        vals = records.env.cache.get_until_miss(records, self)
+        if len(vals) < len(records):
+            remaining = records[len(vals):]
+
+            remaining_new = remaining.filtered(lambda r: not r.id)
+            if remaining_new:
+                # populate the cache for each NewId record
+                [self.__get__(rec, type(rec)) for rec in remaining_new]
+            else:
+                missing = records._browse(records.env, (remaining._ids[0],), tuple(remaining._ids))
+                # trigger the prefetch of all remaining records and store the values in cache
+                self.__get__(missing, type(missing))
+            return vals + self.mapped(remaining)
```

Comment:
I am not fan of a recursive call for such an obviously iterative process.
What about a simple while loop?
```python
vals = records.env.cache.get_until_miss(records, self)
while len(vals) < len(records):
    remaining = records[len(vals):]
    # simplified, simply trigger prefetching from the first remaining record
    self.__get__(first(remaining), type(remaining))
    vals += records.env.cache.get_until_miss(remaining, self)
```